components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -5,68 +5,565 @@
+#include <tuple>
 #include <utility>
 #include <vector>
 
+#include "base/containers/contains.h"
+#include "base/containers/flat_set.h"
 #include "base/functional/bind.h"
+#include "base/logging.h"
 #include "base/memory/scoped_refptr.h"
+#include "base/rand_util.h"
 #include "base/ranges/algorithm.h"
+#include "base/time/time.h"
 #include "brave/components/brave_news/browser/channels_controller.h"
 #include "brave/components/brave_news/browser/feed_fetcher.h"
 #include "brave/components/brave_news/browser/publishers_controller.h"
-#include "brave/components/brave_news/common/brave_news.mojom-forward.h"
+#include "brave/components/brave_news/browser/signal_calculator.h"
 #include "brave/components/brave_news/common/brave_news.mojom.h"
+#include "components/prefs/pref_service.h"
 #include "services/network/public/cpp/shared_url_loader_factory.h"
 
 namespace brave_news {
 
+namespace {
+
+// Minimum number of articles in an inline block.
+constexpr int kBlockInlineMin = 1;
+
+// Maximum number of articles in an inline block.
+constexpr int kBlockInlineMax = 5;
+
+// Ratio of inline articles to discovery articles.
+constexpr double kInlineDiscoveryRatio = 0.25;
+
+// Weighting for unsubscribed sources. Small but non-zero, so we still include
+// them in the feed.
+constexpr double kSourceSubscribedMin = 1e-5;
+
+// Weighting for subscribed sources (either this or |kSourceSubscribedMin| will
+// be applied).
+constexpr double kSourceSubscribedMax = 1;
+
+// Multiplier for unvisited sources. |visit_weighting| is in the range [0, 1]
+// inclusive and will be shifted by this (i.e. [0, 1] ==> [0.2, 1.2]). This
+// ensures we still show unvisited sources in the feed.
+constexpr double kSourceVisitsMin = 0.2;
+
+// Every N hours the popRecency half life will halve. I.e, if this was 24, every
+// day the popularity score will be halved.
+constexpr double kPopRecencyHalfLifeInHours = 18;
+
+// An ArticleInfo is a tuple of (item, signal, weighting). They're stored like
+// this for performance, so we only need to calculate things once.
+using ArticleInfo = std::tuple<mojom::FeedItemMetadataPtr, Signal, double>;
+using ArticleInfos = std::vector<ArticleInfo>;
+
+const Signal* GetSignal(const mojom::FeedItemMetadataPtr& article,
+                        const Signals& signals) {
+  auto it = signals.find(article->url.spec());
+  if (it == signals.end()) {
+    return nullptr;
+  }
+  return &it->second;
+}
+
+double GetPopRecency(const mojom::FeedItemMetadataPtr& article) {
+  auto& publish_time = article->publish_time;
+
+  // TODO(fallaciousreasoning): Use the new popularity field instead.
+  // https://github.com/brave/brave-browser/issues/32173
+  double popularity = article->score == 0 ? 50 : article->score;
+  double multiplier = publish_time > base::Time::Now() - base::Hours(5) ? 2 : 1;
+  auto dt = base::Time::Now() - publish_time;
+
+  return multiplier * popularity *
+         pow(0.5, dt.InHours() / kPopRecencyHalfLifeInHours);
+}
```

Comment:
I'm wondering if we want to extract this into a static function and reuse it from signal calculator. But if this is supposed to be replaced with a server data as you mentioned, please ignore this comment 😄

Reply:
Nice catch! This was duplicated. I just removed the one from the signal calculator - the popularity score has been tacked on to the articles that the server returns, so it doesn't really need to be calculated by the signal calculator :smile: 

We'll just use the one from the FeedV2Builder (and it will use the new value from the server, when it becomes available).